Autopilot/Src/altimeter.cpp

```diff
@@ -0,0 +1,136 @@
+#include "altimeter.hpp"
+
+#include "stm32f7xx_hal.h"
+#include "stm32f7xx_hal_i2c.h"
+
+#include <algorithm>
+
+/***********************************************************************************************************************
+ * Definitions
+ **********************************************************************************************************************/
+
+#define I2C_BUS 1
+#define LEFT_SHIFTED_ALTIMETER_SLAVE_ADDRESS 0xC0 // actual slave address is 0x60. But for whatever reason, the cubeMX drivers don't left shift the address so I have to do that myself
+#define ALT_CTRL_REG1 0x26
+#define ALT_OUT_P_MSB_REG 0x01
+#define STANDBY_MODE_8_OS 0x98
+#define ACTIVE_MODE_8_OS 0x99
+
+#define MEM_SIZE_1_B 1
+
+/***********************************************************************************************************************
+ * Static variables
+ **********************************************************************************************************************/
+
+static bool dataIsNew;
+
+/***********************************************************************************************************************
+ * HAL SPI handle
+ **********************************************************************************************************************/
+
+extern I2C_HandleTypeDef hi2c4;
+
+/***********************************************************************************************************************
+ * Public methods
+ ***********************************************************************:o**********************************************/
+
+Altimeter& MPL3115A2::getInstance()
+{
+    static MPL3115A2 singleton;
+    return singleton;
+}
+
+void MPL3115A2::Begin_Measuring(void)
+{
+    HAL_I2C_Mem_Read(&hi2c4, LEFT_SHIFTED_ALTIMETER_SLAVE_ADDRESS, ALT_OUT_P_MSB_REG, MEM_SIZE_1_B, rawAltimeter.byte, 1, 10);
+}
+
+static float dataLog[100];
+
+void MPL3115A2::GetResult(AltimeterData_t &Data)
+{
+    static int j = 0;
+
+    std::reverse(rawAltimeter.byte, rawAltimeter.byte + 3);
+
+    float tmp = static_cast<float> (rawAltimeter.Q16point4 >> 4);  // the 4 lowest bits of the data register are reserved
+    tmp /= 16.0f;   // conversion from Q16.4 fixed point to float
+    Data.altitude = tmp - altimeterCalibrationFinal;
+
+    Data.isDataNew = dataIsNew;
+    dataIsNew = false;
+}
+
+/***********************************************************************************************************************
+ * Private methods
+ **********************************************************************************************************************/
+
+MPL3115A2::MPL3115A2()
+{
+    HAL_I2C_MspInit(&hi2c4);
+
+    ConfigAltimeter();
+
+    //Calibrate();
```

Comment:
Any reason we are not calibrating?

Reply:
To calibrate, the plane has to be in an "ideal" setting. That means level, not moving, and all this kind of stuff.

So what needs to be done is to put the plane in those ideal conditions, calibrate, and store the calibration values in EEPROM for future use. My plan is to do this once on the WATERLOO airstrip and hardcode the values in the module. It doesent make sense to do this every time the micro is fired up because those ideal conditions aren't always guarantedd.

Reply:
These things aren't as important for the altimeter as they are for something like an IMU, but I'll just do it all one shot.

Reply:
sounds good